x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/DataNodeRequestSender.java

```diff
@@ -322,7 +321,13 @@ private void maybeScheduleRetry(ShardId shardId, boolean receivedData, Exception
                 if (receivedData == false
                     && targetShards.getShard(shardId).remainingNodes.isEmpty()
                     && unwrapFailure(shardId, e) instanceof NoShardAvailableActionException) {
-                    pendingRetries.add(shardId);
+                    if (pendingRetries == null && remainingUnavailableShardResolutionAttempts.decrementAndGet() >= 0) {
+                        pendingRetries = new HashSet<>();
+                        sendingLock.lock();
```

Comment:
I am concerned that we are scattering sendingLock#lock and sendingLock#unlock in two different places. Can we  keep them close?

Reply:
I am not following. They are in the same inner listener, guarding the same structure.

Reply:
@idegtiarenko Sorry, I should have provided more detail. My concern is that we acquire the sending lock in `maybeScheduleRetry` and release it in `onAfter`, which is linked to the status of `pendingRetries`. While the implementation is technically correct, I think we should stick to the simplest lock pattern unless there is a strong reason to do otherwise:

```
lock/tryLock
try {
  ...
} finally {
  unlock
}
```

I think we can follow this lock pattern in the DataNodeRequestSender class.

Reply:
This sounds a lot like the original implementation (within `onAfter`) that was not correct.
Do you see a way how to implement this suggestion while still locking conditionally (only if shard movement is detected) and blocking concurrent requests while it is detected that new shard location resolution is required?

Reply:
As discussed yesterday, I moved retry scheduling to `trySendingRequestsForPendingShards` in https://github.com/elastic/elasticsearch/pull/127464/commits/8a0dcc6a08aa84c2d91923786a4b73c2f20bd9fd